p4-16/psa/examples/psa-example-parser-checksum.p4

```diff
@@ -0,0 +1,232 @@
+    }
+}
+
+parser IngressParserImpl(packet_in buffer,
+                         out headers parsed_hdr,
+                         inout metadata user_meta,
+                         in psa_ingress_parser_input_metadata_t istd,
+                         out psa_parser_output_metadata_t ostd)
+{
+    Checksum<bit<16>>(HashAlgorithm.ones_complement16) ck;
+    state start {
+        buffer.extract(parsed_hdr.ethernet);
+        transition select(parsed_hdr.ethernet.etherType) {
+            0x0800: parse_ipv4;
+            default: accept;
+        }
+    }
+    state parse_ipv4 {
+        buffer.extract(parsed_hdr.ipv4);
+        // TBD: It would be good to enhance this example to
+        // demonstrate checking of IPv4 header checksums for IPv4
+        // headers with options, but this example does not handle such
+        // packets.
+        verify(parsed_hdr.ipv4.ihl == 5, error.UnhandledIPv4Options);
+        // TBD: Update the code below for checking the IPv4 header
+        // checksum with whatever the API we decide upon for a PSA
+        // checksum extern.
+        ck.clear();
+        ck.update({ parsed_hdr.ipv4.version,
+                parsed_hdr.ipv4.ihl,
+                parsed_hdr.ipv4.diffserv,
+                parsed_hdr.ipv4.totalLen,
+                parsed_hdr.ipv4.identification,
+                parsed_hdr.ipv4.flags,
+                parsed_hdr.ipv4.fragOffset,
+                parsed_hdr.ipv4.ttl,
+                parsed_hdr.ipv4.protocol,
+                parsed_hdr.ipv4.hdrChecksum,
+                parsed_hdr.ipv4.srcAddr,
+                parsed_hdr.ipv4.dstAddr });
+        // The verify statement below will cause the parser to enter
+        // the reject state, and thus terminate parsing immediately,
+        // if the IPv4 header checksum is wrong.  It will also record
+        // the error error.BadIPv4HeaderChecksum, which will be
+        // available in a metadata field in the ingress control block.
+        verify(ck.get() == 0, error.BadIPv4HeaderChecksum);
+        transition select(parsed_hdr.ipv4.protocol) {
+            6: parse_tcp;
+            default: accept;
+        }
+    }
+    state parse_tcp {
+        buffer.extract(parsed_hdr.tcp);
+        transition accept;
+    }
+}
+
+control ingress(inout headers hdr,
+                inout metadata user_meta,
+                PacketReplicationEngine pre,
+                in  psa_ingress_input_metadata_t  istd,
+                out psa_ingress_output_metadata_t ostd)
+{
+    Counter<PacketCounter_t, ErrorIndex_t>
+        ((bit<32>) NUM_ERRORS, CounterType_t.packets) parser_error_counts;
+    apply {
+        if (istd.parser_error != error.NoError) {
+            // Example code showing how to count number of times each
+            // kind of parser error was seen.
+            ErrorIndex_t error_idx;
+            convert_error_to_idx(istd.parser_error, error_idx);
+            parser_error_counts.count(error_idx);
+            ingress_drop(ostd);
```

Comment:
This is probably easier to do with an exact match table and a direct counter...

Reply:
Makes sense, and I can change the example to do that.  It simply didn't occur to me while writing the example.  @antoninbas The P4Runtime API should be able to handle exact matches on error type, then?

Reply:
Example changed to use exact match table and direct counter.